### PR TITLE
[ENH] New model API & refactor

### PR DESCRIPTION
**Pull request by mbeyeler**
**INCLUDES NEW API**

Massive re-write of pulse2percept. Major changes:

- [X] New API. The `Simulation` object is gone (this implicitly fixes issue #90). Instead, directly instantiate a model; e.g.
  ```
  >>> import pulse2percept as p2p
  >>> model = p2p.models.AxonMapModel()
  >>> model.build()
  >>> ...etc.
  ```
- [X] New models.
   - [X] Scoreboard model: all phosphenes are Gaussian blobs; Cython back end
   - [X] Axon map model: phosphenes are elongated along nerve fiber bundles; Cython back end
- [X] Major refactor.
   - [X] Submodules are now proper subpackages.
   - [X] The `retina` submodule has been replaced by `models`.
   - [X] The `files` submodule has been replaced by `io`.
   - [X] "Implants" are now subclassed from `ProsthesisSystem`, which combine an `ElectrodeArray` and (optional) a `Stimulus` object. This implicitly fixes issue #74. In the future, we should add video mode vs. direct stimulation mode and VPU functionality.
   - [X] `Stimulus` objects can be built from images (TODO: dicts,pulse trains).
   - [X] New `xarray` objects are used to create N-D arrays with annotations.
- [X] New build process.
   - [X] Compile code and run tests via `Makefile`
   - [X] Use `setuptools` where available
   - [X] Check for NumPy and Cython first
   - [X] Drop Python 2.7 support

### CODE REVIEW

**Comment by mbeyeler** [comment]
Ready for the PR Master

File: pulse2percept/io/image.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        raise ImportError("You do not have scikit-image installed. "
+                          "You can install it via $ pip install scikit-image.")
+
+    # Make sure range of values is valid
+    assert len(valrange) == 2 and valrange[1] > valrange[0]
+
+    isargus1 = isinstance(implant, ArgusI)
+    isargus2 = isinstance(implant, ArgusII)
+    if not isargus1 and not isargus2:
+        raise TypeError("For now, implant must be of type ArgusI or ArgusII.")
```
Comment (arokem) [inline pulse2percept/io/image.py]: Is this still true? `AlphaIMS` will not work?

File: pulse2percept/io/tests/test_video.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+# Import the whole module so we can reload it:
+from pulse2percept.io import video
+from pulse2percept.io import image
+from pulse2percept import stimuli
+from pulse2percept import implants
+
+
+def test_video2stim():
+    reload(image)
```
Comment (arokem) [inline pulse2percept/io/tests/test_video.py]: Why are these being reloaded? Might want a comment here.

File: pulse2percept/io/tests/test_video.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        metadata = video.load_video_metadata('nothing_there.mp4')
+
+    from skvideo import datasets
+    metadata = video.load_video_metadata(datasets.bikes())
+    npt.assert_equal(metadata['@codec_name'], 'h264')
+    npt.assert_equal(metadata['@duration_ts'], '128000')
+    npt.assert_equal(metadata['@r_frame_rate'], '25/1')
+
+    # Trigger an import error
+    with mock.patch.dict("sys.modules", {"skvideo": {}, "skvideo.utils": {}}):
```
Comment (arokem) [inline pulse2percept/io/tests/test_video.py]: Oh! This is the reason for the `reload`?

File: pulse2percept/models/_axon_map.pyx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            dist2 = c_pow(arr[idx, 0] - x, 2) + c_pow(arr[idx, 1] - y, 2)
+            gauss[idx] = c_exp(-dist2 / (2.0 * c_pow(tau, 2)))
+    return np.asarray(gauss)
+
+
+cpdef fast_jansonius(double[:] rho, double phi0, double beta_s, double beta_i):
+    cdef double b, c, deg2rad, rho_min, tmp_phi, tmp_rho
+    cdef np.intp_t idx
+
+    deg2rad = 3.14159265358979323846 / 180.0
```
Comment (arokem) [inline pulse2percept/models/_axon_map.pyx]: Implement as module-wide constant?

File: pulse2percept/models/_axon_map.pyx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        for i_stim in range(n_stim):
+            for i_ax in range(n_ax):
+                r2 = c_pow(axon[i_ax, 0] - xel[i_stim], 2)
+                r2 += c_pow(axon[i_ax, 1] - yel[i_stim], 2)
+                gauss = c_exp(-r2 / (2.0 * c_pow(rho, 2)))
+                act[i_ax] += stim[i_stim] * axon[i_ax, 2] * gauss
+    bright = c_max(act)
+    if bright < th:
+        bright = 0
+    return bright
```
Comment (arokem) [inline pulse2percept/models/_axon_map.pyx]: I noticed that all the functions here are called `fast_XXX`. Maybe we can drop that prefix?

File: pulse2percept/models/base.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+            class MyModel(BaseModel):
+
+                def __init__(self, **kwargs):
+                    super(MyModel, self).__init__(self, **kwargs)
+                    self.newvar = 0
+
+        This is the only place where you can add new class attributes. Trying
+        to set `self.someothervar` outside the constructor will raise an
+        AttributeError. Of course, you can always set ``self.newvar = None`` in
```
Comment (arokem) [inline pulse2percept/models/base.py]: I wonder whether we want to implement a custom error for this?

File: pulse2percept/models/watson.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import numpy as np
+import scipy.stats as spst
+
+from ..utils import cart2pol, pol2cart
+
+
+class WatsonConversionMixin(object):
+    """Converts dva to retinal coords using Watson (2014)"""
```
Comment (arokem) [inline pulse2percept/models/watson.py]: Could we have a more complete reference here and below?

File: setup.py [hunk 8]
```diff
@@ -1,37 +1,304 @@
+#! /usr/bin/env python
+#
+# Copyright (C) 2007-2009 Cournapeau David <cournape@gmail.com>
```
Comment (arokem) [inline setup.py]: I am not sure we want to leave the sklearn copyright notice here.

**Pull request merged**